File: Client/AuthService.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Net.NetworkInformation;
using KolibSoft.AuthStore.Core;
using KolibSoft.AuthStore.Core.Abstractions;
using KolibSoft.Catalogue.Core;
using KolibSoft.Catalogue.Core.Utils;

namespace KolibSoft.AuthStore.Client;

public class AuthService : IAuthConnector
{

    public HttpClient HttpClient { get; }
    public string Uri { get; }

    public virtual bool Available => NetworkInterface.GetAllNetworkInterfaces().Any(x => x.OperationalStatus == OperationalStatus.Up && x.NetworkInterfaceType != NetworkInterfaceType.Loopback);

    public virtual async Task<Result<AuthModel?>> AccessAsync(LoginModel login)
    {
        var uri = $"{Uri}";
        var response = await HttpClient.PostAsJsonAsync(uri, login);
        var result = await response.HandleResult<AuthModel?>();
        return result;
    }

    public virtual async Task<Result<AuthModel?>> RefreshAsync(Guid id, string refreshToken)
    {
        var uri = $"{Uri}/{id}";
        var request = new HttpRequestMessage(HttpMethod.Get, uri);
        request.Headers.Authorization = new AuthenticationHeaderValue("bearer", refreshToken);
        var response = await HttpClient.SendAsync(request);
        var result = await response.HandleResult<AuthModel?>();
        return result;
    }

    public AuthService(HttpClient httpClient, string uri)
    {
        HttpClient = httpClient;
        Uri = uri;
    }

}

File: docs/lib/services/auth_service.js
import { AuthConnector } from "../abstractions/auth_connector.js";
import { AuthModel } from "../models/auth_model.js";
import { Result } from "../modules/catalogue.js";

/**
 * @implements {AuthConnector}
 */
class AuthService {

    /** @type {(input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>} */
    fetch;

    /** @type {string} */
    uri;

    /** @returns {boolean} */
    get available() { return navigator.onLine; }

    /**
     * @param {LoginModel} login 
     * @returns {Promise<Result<AuthModel>>}
     */
    async accessAsync(login) {
        let uri = `${this.uri}`;
        let response = await this.fetch(uri, {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(login)
        });
        let result = await Catalogue.ResultUtils.handleResult(response);
        if (result.data) result.data = new AuthModel(result.data);
        return result;
    }

    /**
     * @param {string} id 
     * @param {string} refreshToken 
     * @returns {Promise<Result<AuthModel>>}
     */
    async refreshAsync(id, refreshToken) {
        let uri = `${this.uri}/${id}`;
        let response = await this.fetch(uri, {
            method: "GET",
            headers: {
                "Authorization": `bearer ${refreshToken}`
            }
        });
        let result = await Catalogue.ResultUtils.handleResult(response);
        if (result.data) result.data = new AuthModel(result.data);
        return result;
    }

    /**
     * @param {{(input: RequestInfo | URL, init?: RequestInit | undefined) => Promise<Response>}} fetch 
     * @param {string} uri 
     */
    constructor(fetch, uri) {
        this.fetch = fetch;
        this.uri = uri;
    }

}

export {
    AuthService
}